Enhancement: add --sort-by flag for images command

WeiZhang555 opened the pull request:
Part of #25609

Allow user to customize their favourite list order, user can choose to
sort images by e.g. "size" in "asc"/"des" order.

Usage: --sort-by field[:order]

Supported fields: repository, tag, digest, createdat, id, size

order can be either "asc" or "desc", which means sort in ascendant/desendant
orders, default is "asc" if not specified.

Sample Usage: "--sort-by size:desc"

Signed-off-by: Zhang Wei zhangwei555@huawei.com

This is more like a prototype, I'd like to hear more feedbacks before I started to write test cases. :smile:

WeiZhang555 committed:
Enhancement: add --sort-by flag for images command

Allow user to customize their favourite list order, user can choose to
sort images by e.g. "size" in "asc"/"des" order.

Usage: --sort-by field[:order]

Supported fields:
* "repo": sort by repository:tags.
* "created": sort by created time.
* "size": sort by image size.

order can be either "asc" or "des", which means sort in ascent/descent
orders, default is "asc" if not specified.

Sample Usage: "--sort-by size:des"

Signed-off-by: Zhang Wei <zhangwei555@huawei.com>

yongtang commented:
I am thinking maybe it makes sense to have a fallback order (in addition to the order specified by the user). For example, if `--sort-by size` is specified and two images have the same name, then the fallback order could be the timestamp (or something else as long as we could establish an predictable order).

WeiZhang555 commented:
@yongtang predicatable order is good idea, but user always has their own preference, so it's hard to define a good enough predefined order for user. Besides, I'm worried that it will make the implementation complicated and messy. 
In current implementation, if user didn't specify any "--sort-by" flag, default list will follow the order from daemon; if user specify one, then first order is "custom", fallback order will keep same with daemon.

cpuguy83 commented:
This is fine for me.
Only comment I'd make is if we're really going to add sorting, doing it by multiple columns would be nice.... not a requirement for adding IMO.

WeiZhang555 commented:
So what's other guys' opinions? I hope to get more feedback from all :smile: 
ping @vdemeester @tonistiigi

WeiZhang555 commented:
Is this ready to move forward to code review? @cpuguy83 @tonistiigi @vdemeester

tonistiigi commented:
I'm not sure it is worth it. Seems like not so important feature that needs a lot of code. It will be inconsistent with other list views that don't support it and even if we add it to everywhere I think we can easily break it when new fields are added. Also user has little way to know that columns are supported without going through the docs. Maybe I would think differently if we'd use some kind of component for all the list views that would provide this capability for free(and also fix `--format {{json .}}` for example).

cpuguy83 commented:
A generic/reflect-based sort could be used for all types and doesn't matter if new fields are added as long as he handle the base type on the field.
It's all on the client, so not too worried about reflect usage for this.

WeiZhang555 commented:
@tonistiigi I think @cpuguy83 provides a nice suggestion for solving the "new field" problem, I can try to bring it in reality.

>Also user has little way to know that columns are supported without going through the docs.

Actualy this is not a new problem, look at the `--filters` everywhere, I can never remember all of them and everytime I have to `man docker-xxx` for its name. For example:

```
# man docker-ps
......
       -f, --filter=[]
          Filter output based on these conditions:
          - exited=<int> an exit code of <int>
          - label=<key> or label=<key>=<value>
          - status=(created|restarting|running|paused|exited|dead)
          - name=<string> a container's name
          - id=<ID> a container's ID
          - before=(<container-name>|<container-id>)
          - since=(<container-name>|<container-id>)
          - ancestor=(<image-name>[:tag]|<image-id>| ⟨image@digest⟩) - containers created from an image or a descendant.
          - volume=(<volume-name>|<mount-point-destination>)
          - network=(<network-name>|<network-id>) - containers connected to the provided network
```

Is there really one person who can always find the right filter name without searching the docs?

WeiZhang555 commented:
@cpuguy83 after thinking more about your idea of generic/reflect-based sort methods, I found that it may not be very appropriate for image sorting, because the fields(sort keys) aren't all fetched from a single structure, also we may need some kinds of union keys, for example: "--sort-by repo" means sorting image by its "repoName"+"tags".

Considering these, I think a generic implementation brings little benefits and I decide not to do it. So that's it, what's your idea?

thaJeztah commented:
We're generally OK with this, but share @tonistiigi's concerns about maintenance; @mlaventure is going to think if a better implementation is possible

WeiZhang555 commented:
@mlaventure Are you still working on this? Any progress?

mlaventure committed:
Add a Generic Struct Sorter

Signed-off-by: Kenfe-Mickael Laventure <mickael.laventure@gmail.com>
Signed-off-by: Zhang Wei <zhangwei555@huawei.com>

mlaventure commented:
@WeiZhang555 sorry, I got distracted by other things :sweat_smile: 

I cooked up something up between yesterday and now (had to get familiar with the `reflect` package first).

You can have a look [here](https://github.com/mlaventure/docker/commit/6e2bb0d8fa41cff628d3fa23c6536919a1c14e25), it's based on your PR branch.

cc @cpuguy83 @tonistiigi

WeiZhang555 commented:
@mlaventure Cool, I just checked the codes, you implemented multi-colume sorting, that's really cool.

I'll paste your commit here and let Janky test it. In general, I think it's OK, but seems too complicated, not sure if we can simplify it a little bit.

thaJeztah commented:
@WeiZhang555 looks like there's a gofmt issue;

```
07:10:25 ---> Making bundle: validate-gofmt (in bundles/1.13.0-dev/validate-gofmt)
07:10:25 These files are not properly gofmt'd:
07:10:25  - cli/command/formatter/sorter.go
07:10:25 
07:10:25 Please reformat the above files using "gofmt -s -w" and commit the result.
```

WeiZhang555 commented:
@thaJeztah Yup, already fixed some tiny test case failure, let's see how it works now. :smile:

vdemeester commented:
@WeiZhang555 

```
09:28:33 ----------------------------------------------------------------------
09:28:33 FAIL: docker_cli_images_test.go:47: DockerSuite.TestImagesOrderedByCreationDate
09:28:33 
09:28:33 docker_cli_images_test.go:65:
09:28:33     c.Assert(imgs[0], checker.Equals, id3, check.Commentf("First image must be %s, got %s", id3, imgs[0]))
09:28:33 ... obtained string = "sha256:11f64303f0f7ffdc71f001788132bca5346831939a956e3e975c93267d89a16d"
09:28:33 ... expected string = "sha256:39b831d802b066eacc9656595fe08bfc700d84b4d4df16cf03bd087f543a29d8"
09:28:33 ... First image must be sha256:39b831d802b066eacc9656595fe08bfc700d84b4d4df16cf03bd087f543a29d8, got sha256:11f64303f0f7ffdc71f001788132bca5346831939a956e3e975c93267d89a16d
09:28:33 
09:28:33 
```

WeiZhang555 commented:
@vdemeester All green now :wink:

thaJeztah commented:
ping @cpuguy83 when you get back from holidays, can you have a look at the reflect part, and say what you think?

design-wise, still looks good to me

cpuguy83 commented:
Oh sure, I suggest reflect and now you make me actually review it :)

thaJeztah commented:
@cpuguy83 well, at least verify this is what you had in mind :-)

cpuguy83 commented on cli/command/formatter/sorter.go:
```diff
@@ -0,0 +1,289 @@
+			nt = nt.Elem()
+		}
+		if nt.Kind() == reflect.Struct {
+			ok = findFieldIndex(nt, name, index)
```
This seems a bit much. Our list views should be simple, i.e. a struct with simple fields.

WeiZhang555 replied:
This part is written by @mlaventure , but I agree with his implementation, I think we'd better keep this, as the image related structures are complicated. 
Right now, it works well for me, I can get what I want.

mlaventure replied:
@WeiZhang555 is correct regarding this being needed for the image struct.

Initially, I was only looking at the first layer, but indeed the image struct is more complicated which forced me to go that way.

cpuguy83 commented:
It seems like it's ok, but some tests sorting the various different `reflect.Kind` types.

mlaventure commented:
@WeiZhang555 wrote some quick unit tests to address @cpuguy83 last comment: https://gist.github.com/mlaventure/61f7de694bcd4a81e5d1303076df30a3

WeiZhang555 commented:
@mlaventure Sorry, I missed your notifications :joy:  Unit test case added :coffee:

thaJeztah commented on cli/command/image/list.go:
```diff
@@ -46,6 +47,7 @@ func NewImagesCommand(dockerCli *command.DockerCli) *cobra.Command {
 	flags.BoolVar(&opts.showDigests, "digests", false, "Show digests")
 	flags.StringVar(&opts.format, "format", "", "Pretty-print images using a Go template")
 	flags.VarP(&opts.filter, "filter", "f", "Filter output based on conditions provided")
+	flags.StringVar(&opts.sortby, "sort-by", "", "Customize list order, sort by field[:order]")
```
This should mention that it takes a comma-separated list.

AkihiroSuda replied:
Can you please add `cli/command/image/list_test.go`?

WeiZhang555 replied:
I'll do it :smile:

WeiZhang555 replied:
Added some unit tests for sort package and one integration test instead. :smile:

thaJeztah commented:
Overall design looks good, some things in the implementation;

- It's hard to find the right fields (e.g. `Name`, `name`, `Repository` ?)
- the error message for an unknown field is confusing `field Name was not found in struct`
- we suggest to use the same "placeholders" for sorting, as can be used for `--format`, and should be visible when `--format '{{ json . }}'`

vdemeester commented:
Also see https://github.com/docker/docker/pull/30193#issuecomment-273042092

mlaventure commented:
@vdemeester you can specify as many sort key as you want :)

ping @WeiZhang555, do you think you can address @thaJeztah comments?

WeiZhang555 commented:
@mlaventure Sorry for the delay, a little busy these days, also I was working on another PR: https://github.com/docker/docker/pull/25361 , hope to get it in :smile: 

Back to this one: @thaJeztah 

> * It's hard to find the right fields (e.g. Name, name, Repository ?) 
Maybe we can add fields to docs as a reference? 

> * the error message for an unknown field is confusing field Name was not found in struct
I think I can add an error wrapper to make error message more friendly and meaningful

> * we suggest to use the same "placeholders" for sorting, as can be used for --format, and should be visible when --format '{{ json . }}'
A "placeholder" for sorting? How does it work? I didn't get it, sorry...What will it be looking like?

AkihiroSuda commented:
I wonder the support for "placeholder"s has been already supported via `reflect`. (e.g. `"Created"`)


As mentioned above, we need an automated test to confirm the behavior about "placeholder"s.

WeiZhang555 commented:
> I wonder the support for "placeholder"s has been already supported via reflect. (e.g. "Created")

Yep, I think so. But let me test, also add a test case for it. :smile:

WeiZhang555 committed:
Refactor sorter

Optimize implementation of sorter, also fix some bugs. One important
enhancement is printing hint to user when user specify wrong `--sort-by`
value, e.g.

```
$ docker images --sort-by a
can't sort by "a", it must be in
[trunc,Containers,ParentID,SharedSize,repo,digest,Created,ID,Size,VirtualSize,tag]
```

Signed-off-by: Zhang Wei <zhangwei555@huawei.com>

WeiZhang555 commented:
@thaJeztah I made some changes according to your comments:

> It's hard to find the right fields (e.g. Name, name, Repository ?)

Now it will print hints to user when user specify wrong fields, for example:
```
$ docker images --sort-by a
can't sort by "a", it must be in [digest,trunc,Containers,ParentID,SharedSize,repo,tag,Created,ID,Size,VirtualSize]
```
What do you think of this format?

> the error message for an unknown field is confusing `field Name was not found in struct`

See lines above.

> we suggest to use the same "placeholders" for sorting, as can be used for --format, and should be visible when --format '{{ json . }}'

Currently no, sorter is using the struct field name. Maybe we should add a white list to map our existing placeholder to the struct name? For example, "CreatedAt"(format placeholder) map to "Created"(sorter)?

@mlaventure I fixed something I thought wrong, also try hard to simplied the implementation. Could you please help review?

WeiZhang555 commented:
@mlaventure Also, I found that you `t.FieldByNameFunc(match)` doesn't work as expected,  these are two examples: https://play.golang.org/p/aZJPABCUKv doesn't work, but https://play.golang.org/p/_F51etTBe8 works. Unit test case `TestFindFieldIndex` can also verify this. I didn't figure out why :crying_cat_face:

mlaventure commented:
@WeiZhang555 it fails because the function returns `true` twice. This example works: https://play.golang.org/p/xYoUzz_HPo

Since it should only match public fields, this example is better: https://play.golang.org/p/p6pyZ24497

mlaventure commented on cli/command/formatter/sorter.go:
```diff
@@ -7,11 +7,20 @@ import (
-	ordering []string
-	indexes  [][]int
+	data []interface{}
+	bys  map[string]orderBy
```
A map doesn't preserve the order passed by the user. This has to be a slice.

WeiZhang555 replied:
Ooops, you are definitely right. I'll modify~

WeiZhang555 commented:
Added a new enhancement:
```
    Add `sortby` tag for sortable field
    
    This is enhancement to sort implementation, we use a `sortby` tag to
    identify a field sortable, for example
    
    ```
    type s struct{
            elem1 string `sortby:"Name"`
            elem2 string
    }
    ```
    
    Means you can speficy `--sort-by Name` to sort list of structures based
    on `elem1` field. Fields without `sortby` tag aren't sortable.
```

If you specify wrong fields:

```
$ docker images --sort-by a
can't sort by "a", it must be in [ID,Size,Repository,Tag,Digest,CreatedAt]
```

Here there's a whitelist calculated based on `sortby` tag, I'm doing this because I don't think all fields in image struct are appropriate to sort such as `ParentID` or `trunc`, also the list is consistent with `--format` placeholder now :smile: 

ping @mlaventure @thaJeztah

mlaventure reviewed (changes requested):
This won't work because the tags aren't generated by the swagger (hence the janky failure). I'm not sure if it's possible to add arbitrary tag to a field with swagger though (cc @dnephin)

Also, could we have the sort fields being case insensitive? :)

dnephin commented:
It's not supported by go-swagger.

All this reflection seems a bit unnecessary. I think it could just be a single switch/case statement. The client doesn't really need to be aware of which fields are sortable.

WeiZhang555 commented:
@dnephin 

> All this reflection seems a bit unnecessary. I think it could just be a single switch/case statement. The client doesn't really need to be aware of which fields are sortable.

I'm doing this because 1. not every field in a struct is meaningful for the `sort` feature, as the `sorter` is a common processor, it must have a efficient filter mechanism to ignore unmeaningful field.  2. Some field name is not very formal, like `repo`, `tag`, it's different from printout of `docker images`, and also the placeholder is better to be consistent with `--format`(e.g. `Repository`, `Tag`), we must have a method to rename the filed for sorter value.

I'll try more, if swagger can't support, I could use some `map` structure as whitelist, but prefer the reflect tag, sad... :cry: 

> Also, could we have the sort fields being case insensitive? :)

OK, no problem. Update is coming...

dnephin commented:
A map structure or switch/case will also allow for multiple aliases for sorting.

WeiZhang555 committed:
Add whitelist support for sorter

Add whitelist support so we can filter the unwanted fields, and give
the fields new name for sorting.

Now image `--sort-by` flag can take same placeholders from `--format`
flag.

Signed-off-by: Zhang Wei <zhangwei555@huawei.com>

WeiZhang555 commented:
Just refactored codes to use map as whitelist, also the "case insensitive issue" is addressed. @mlaventure 

Please help review :smile:

mlaventure commented on cli/command/formatter/sorter.go:
```diff
@@ -0,0 +1,311 @@
+			switch parts[1] {
+			case descendantKey:
+				dir = descendantKey
+			case ascendantKey, "": // if default, keep as ascendantKey
```
nit: can you put this under the case for readability

mlaventure commented on cli/command/formatter/sorter.go:
```diff
@@ -0,0 +1,311 @@
+			case descendantKey:
+				dir = descendantKey
+			case ascendantKey, "": // if default, keep as ascendantKey
+			default: // doesn't support key other than "asc","dsc",""
```
nit: can you put this under `default:` for readability

mlaventure commented on cli/command/formatter/sorter.go:
```diff
@@ -0,0 +1,311 @@
+			default: // doesn't support key other than "asc","dsc",""
+				return gs, fmt.Errorf("sort order %q not supported", parts[1])
+			}
+		case 1: // sort order not specified, keep default
```
nit: can you put this under the case for readability

mlaventure commented on cli/command/formatter/sorter.go:
```diff
@@ -0,0 +1,311 @@
+		}
+
+		switch fieldType.Kind() {
+		case reflect.String, reflect.Bool, reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64, reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64, reflect.Float32, reflect.Float64:
```
Can you split those for readability? You can go to the next line after a comma.

May separate them by equivalent type?

mlaventure commented on cli/command/formatter/sorter.go:
```diff
@@ -0,0 +1,311 @@
+				// we don't support sorting by ordinary struct type, continue to skip
+				continue
+			}
+		default: //ignore other types
```
nit: can you put this under the `default:` for readability

mlaventure reviewed (changes requested):
a few nits, but LGTM

WeiZhang555 commented:
@mlaventure Updated.

mlaventure reviewed (approved):
LGTM

cpuguy83 commented:
@dnephin I moved this to code review, are you ok with this?

dnephin commented:
Ya, this works for me. It still uses reflect, but I like that mapping instead of struct tags.

WeiZhang555 commented:
@mlaventure Who else should I ping?

htc2u commented:
Ping to htc2u@htc2social.blog and elsonchow3833@gmail.com  in htc2u.com

发自我的 iPhone

> 在 2017年2月13日，上午9:46，zhangwei_cs <notifications@github.com> 写道：
> 
> @mlaventure Who else should I ping?
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

AkihiroSuda commented on cli/command/image/list.go:
```diff
@@ -46,6 +47,7 @@ func NewImagesCommand(dockerCli *command.DockerCli) *cobra.Command {
 	flags.BoolVar(&opts.showDigests, "digests", false, "Show digests")
 	flags.StringVar(&opts.format, "format", "", "Pretty-print images using a Go template")
 	flags.VarP(&opts.filter, "filter", "f", "Filter output based on conditions provided")
+	flags.StringVar(&opts.sortby, "sort-by", "", "Customize list order, sort by field[:order]. E.g. \"ID:asc,Created:dsc\"")
```
nit: E.g. -> e.g.

htc2u replied:
nit: E.g. -> e.g.

AkihiroSuda commented on cli/command/formatter/sorter.go:
```diff
@@ -0,0 +1,319 @@
+				// if default, keep as ascendantKey
+			default:
+				// doesn't support key other than "asc","dsc",""
+				return gs, fmt.Errorf("sort order %q not supported", parts[1])
```
Can you please print supported keys ("asc", "dsc") here for UX?

WeiZhang555 replied:
OK

AkihiroSuda commented on cli/command/formatter/sorter.go:
```diff
@@ -0,0 +1,319 @@
+
+const (
+	ascendantKey  = "asc"
+	descendantKey = "dsc"
```
I wonder `desc` is widely used abbreviation.

Google result:
- "sort asc dsc": 350,000  hits
- "sort asc desc": 15,100,000 hits

WeiZhang555 replied:
I can't be sure, but looks like you are right:

https://www.allacronyms.com/descendant/abbreviated
http://www.rootsweb.ancestry.com/~rigenweb/abbrev.html

I'll modify it to desc

htc2u commented:
Pull request

AkihiroSuda commented on cli/command/formatter/sorter.go:
```diff
@@ -0,0 +1,319 @@
+			for k := range sortableFields {
+				keys = append(keys, k)
+			}
+			return gs, fmt.Errorf("can't sort by %q, it must be in [%s]", by, strings.Join(keys, ","))
```
I found that keys are always `ToLower()`-ed. It would be better if the original casing can be kept, but it can be another PR.

WeiZhang555 commented:
@AkihiroSuda All your comments are addressed :-)

WeiZhang555 commented:
Weird, why `Windows` tests can't recognize `--sort-by` flag?

```
 17:51:13 ----------------------------------------------------------------------
17:51:13 FAIL: docker_cli_images_test.go:367: DockerSuite.TestImageSortBy
17:51:13 
17:51:13 docker_cli_images_test.go:372:
17:51:13     imagesOut, _ := dockerCmd(c, "images", "--sort-by", "Repository:asc,tag:desc", "name*")
17:51:13 docker_utils_test.go:247:
17:51:13     c.Assert(result, icmd.Matches, icmd.Success)
17:51:13 ... result *cmd.Result = &cmd.Result{Cmd:(*exec.Cmd)(0xc042091ce0), ExitCode:125, Error:(*exec.ExitError)(0xc043ba3e20), Timeout:false, outBuffer:(*cmd.lockedBuffer)(0xc042095b00), errBuffer:(*cmd.lockedBuffer)(0xc042095b90)} ("\nCommand: C:\\Windows\\system32\\docker.exe images --sort-by Repository:asc,tag:desc name*\nExitCode: 125, Error: exit status 125\nStdout: \nStderr: unknown flag: --sort-by\nSee 'docker images --help'.\n\n")
17:51:13 ... expected cmd.Expected = cmd.Expected{ExitCode:0, Timeout:false, Error:"", Out:"", Err:""}
17:51:13 ... 
17:51:13 Command: C:\Windows\system32\docker.exe images --sort-by Repository:asc,tag:desc name*
17:51:13 ExitCode: 125, Error: exit status 125
17:51:13 Stdout: 
17:51:13 Stderr: unknown flag: --sort-by
17:51:13 See 'docker images --help'.
17:51:13 
17:51:13 
17:51:13 Failures:
17:51:13 ExitCode was 125 expected 0
17:51:13 Expected no error
17:51:13 
17:51:13 
17:51:13 
17:51:13 ----------------------------------------------------------------------
```

@jhowardmsft Any idea on this?

jhowardmsft commented:
@WeiZhang555 Nothing obvious I can see from the change - there's nothing _unix.go/_Linux.go specific. I restarted CI, doubtful as I am it will make a difference. I'd have to debug it to work out why when I get time if it fails next time, but I don't have any idea right now.

vdemeester commented on cli/command/image/list.go:
```diff
@@ -46,6 +47,7 @@ func NewImagesCommand(dockerCli *command.DockerCli) *cobra.Command {
 	flags.BoolVar(&opts.showDigests, "digests", false, "Show digests")
 	flags.StringVar(&opts.format, "format", "", "Pretty-print images using a Go template")
 	flags.VarP(&opts.filter, "filter", "f", "Filter output based on conditions provided")
+	flags.StringVar(&opts.sortby, "sort-by", "", "Customize list order, sort by field[:order]. e.g. \"ID:asc,Created:desc\"")
```
This help is not up-to-date 👼

```
can't sort by "created", it must be in [repository,tag,digest,createdat,id,size]
```

WeiZhang555 replied:
Ahh, I'm sorry...I'll do a update now.

vdemeester reviewed (changes requested):
A small comment, otherwise looks good :angel: 
/cc @dnephin @thaJeztah

WeiZhang555 commented:
@vdemeester Fixed. Let's see if `Windows CI` works now (quite weird problem :confused: )

vdemeester reviewed (approved):
LGTM 🐸

vdemeester commented:
@WeiZhang555 hum the windows failure looks legit though

```
02:52:56 ----------------------------------------------------------------------
02:52:56 FAIL: docker_cli_images_test.go:367: DockerSuite.TestImageSortBy
02:52:56 
02:52:56 docker_cli_images_test.go:372:
02:52:56     imagesOut, _ := dockerCmd(c, "images", "--sort-by", "Repository:asc,tag:desc", "name*")
02:52:56 docker_utils_test.go:247:
02:52:56     c.Assert(result, icmd.Matches, icmd.Success)
02:52:56 ... result *cmd.Result = &cmd.Result{Cmd:(*exec.Cmd)(0xc04208f600), ExitCode:125, Error:(*exec.ExitError)(0xc0422b1a40), Timeout:false, outBuffer:(*cmd.lockedBuffer)(0xc0423366c0), errBuffer:(*cmd.lockedBuffer)(0xc042336750)} ("\nCommand: C:\\Windows\\system32\\docker.exe images --sort-by Repository:asc,tag:desc name*\nExitCode: 125, Error: exit status 125\nStdout: \nStderr: unknown flag: --sort-by\nSee 'docker images --help'.\n\n")
02:52:56 ... expected cmd.Expected = cmd.Expected{ExitCode:0, Timeout:false, Error:"", Out:"", Err:""}
02:52:56 ... 
02:52:56 Command: C:\Windows\system32\docker.exe images --sort-by Repository:asc,tag:desc name*
02:52:56 ExitCode: 125, Error: exit status 125
02:52:56 Stdout: 
02:52:56 Stderr: unknown flag: --sort-by
02:52:56 See 'docker images --help'.
02:52:56 
02:52:56 
02:52:56 Failures:
02:52:56 ExitCode was 125 expected 0
02:52:56 Expected no error
02:52:56 
02:52:56 
02:52:56 
02:52:56 ----------------------------------------------------------------------
```

AkihiroSuda reviewed (approved):
LGTM, moving to docs review

WeiZhang555 commented:
@vdemeester This is also confusing me, why it can't recognize `--sort-by` flag? Could this be a test server problem? Maybe the client is old which doesn't own the new flag?

vdemeester commented:
/cc @johnstep @jhowardmsft 👼 🙏

jhowardmsft commented:
I think I know what it is. Will investigate shortly.

jhowardmsft commented:
Yes, I know what it is, have a fix for CI, but I've lost write access to Jenkins to be able to do the update as of yesterday. Ping @wfarner

jhowardmsft commented:
Jenkins has been updated with my fix and I restarted it. Hopefully it'll pass now.

WeiZhang555 commented:
@jhowardmsft Yes, it works! Thank you for your help!  :smiley: 

I'll update docs today.

vdemeester reviewed (changes requested):
@WeiZhang555 could you update the man pages and `docs/reference` for the `images` command to show the new flag (with maybe an example) 👼.

/cc @thaJeztah

WeiZhang555 committed:
Add docs for `images --sort-by` flag

Signed-off-by: Zhang Wei <zhangwei555@huawei.com>

WeiZhang555 commented:
Dos added (&rebased) :smile: 
/cc @thaJeztah

AkihiroSuda commented on docs/reference/commandline/images.md:
```diff
@@ -343,3 +344,39 @@ b6fa739cedf5        committ                   latest
+
+## Customizing display orders
+
+To customize list orders, you can use `--sort-by` option, this option takes
```
I think this sentence should be split to multiple sentences.

e.g.
```
To customize list orders, you can use `--sort-by` option, this option takes format of `field[:order]`.
`field` should be a case insensitive string among "repository", "tag", "digest", "createdat", "id", and "size".
`order` should be either `asc`(ascendant) or `desc`(descendant).
```

AkihiroSuda commented on docs/reference/commandline/images.md:
```diff
@@ -343,3 +344,39 @@ b6fa739cedf5        committ                   latest
+format of `field[:order]`, in which `field` should be a case insensitive string
+among "repository", "tag", "digest", "createdat", "id", and "size", `order`
+should be either `asc`(ascendant) or `desc`(descendant). Meanings of each `field`
+value has been described in [`Formatting`](#formatting) section.
```
nit: has been -> are

WeiZhang555 commented:
@AkihiroSuda Updated as per your suggestion, thank! :smiley:

vdemeester commented on docs/reference/commandline/images.md:
```diff
@@ -33,6 +33,7 @@ Options:
       --help            Print usage
       --no-trunc        Don't truncate output
   -q, --quiet           Only show numeric IDs
+      --sort-by string  Customize list order, sort by field[:order]. e.g. "ID:asc,CreatedAt:desc"
```
I wonder if we should use always lowercase on the docs (and in the help), @thaJeztah wdyt ?

WeiZhang555 replied:
I'm fine with lowercase everywhere, as this flag takes case insensitive string.

vdemeester commented on docs/reference/commandline/images.md:
```diff
@@ -343,3 +344,39 @@ b6fa739cedf5        committ                   latest
+of `field[:order]`. `field` should be a case insensitive string among "repository",
+"tag", "digest", "createdat", "id" and "size". `order` should be either `asc`(ascendant)
+or `desc`(descendant).
+Meanings of each `field` value are described in [`Formatting`](#formatting) section.
```
`The meaning of each field is described in …`

vdemeester commented on docs/reference/commandline/images.md:
```diff
@@ -343,3 +344,39 @@ b6fa739cedf5        committ                   latest
+or `desc`(descendant).
+Meanings of each `field` value are described in [`Formatting`](#formatting) section.
+
+If you specify wrong `field` or `order` string, it will output the right options
```
`… specify a wrong …`

vdemeester commented on man/src/image/ls.md:
```diff
@@ -115,3 +115,29 @@ Listing just the shortened image IDs. This can be useful for some automated
+"tag", "digest", "createdat", "id" and "size". `order` should be either `asc`(ascendant)
+or `desc`(descendant).
+
+If you specify wrong `field` or `order` string, it will output the right options
```
Same here.

vdemeester commented on man/src/image/ls.md:
```diff
@@ -115,3 +115,29 @@ Listing just the shortened image IDs. This can be useful for some automated
+    docker images --sort-by invalid
+    can't sort by "invalid", it must be in [repository,tag,digest,createdat,id,size]
+
+If you hope to sort images with their repository name in ascendant order, then sort
```
`To sort images by repository name in ascendant order, then by tags in […]`

vdemeester reviewed (changes requested):
/cc @thaJeztah @mstanleyjones

WeiZhang555 commented:
@vdemeester Updated.

vdemeester reviewed (approved):
LGTM 🐸
/cc @thaJeztah @mstanleyjones

thaJeztah commented on cli/command/formatter/image.go:
```diff
@@ -174,10 +178,34 @@ func imageFormat(ctx ImageContext, images []types.ImageSummary, format func(subC
+		return err
+	}
+
+	sort.Sort(sorter)
```
Looks like we're now always sorting, but before this change, we didn't? I think we should only create a sorter if `ctx.SortBy` is not empty / `bys` is not empty, and otherwise print the results as-is. The results are already sorted in the daemon; https://github.com/docker/docker/blob/254fc83cba90ed79c78f4cb0cb33aeeaff492798/daemon/images.go#L230

thaJeztah commented on cli/command/formatter/sorter.go:
```diff
@@ -0,0 +1,321 @@
+		return gs, nil
+	}
+	if len(bys) == 0 {
+		return gs, fmt.Errorf("bys cannot be empty")
```
If we only create a sorter if bys is not empty, this check is not needed.

Also, in which situation would we get an empty list (and is it an actual _error_?) Could probably be just;

```go
if len(data) == 0 || len(bys) == 0 {
    return gs, nil
}
```

thaJeztah commented on cli/command/formatter/sorter.go:
```diff
@@ -0,0 +1,321 @@
+		return gs, fmt.Errorf("bys cannot be empty")
+	}
+	if err := validateType(data); err != nil {
+		return gs, err
```
```go
return nil, err
```

thaJeztah replied:
Oh ignore, that won't work :blush:

thaJeztah commented on cli/command/formatter/sorter.go:
```diff
@@ -0,0 +1,321 @@
+	// scan and get the sortable fields
+	sortableFields := make(map[string]fieldIndex)
+	if err := getSortableFields(reflect.TypeOf(data[0]), whitelist, nil, sortableFields); err != nil {
+		return gs, fmt.Errorf("can't get sortable fields: %v", err)
```
Also, should be

```go
return nil, fmt.Errorf(".....")
```

mstanleyjones replied:
Try not to use contractions. How about "No sortable fields found"?

WeiZhang555 replied:
And we still need to print the error, right? In case we lost something.

thaJeztah commented on cli/command/formatter/sorter.go:
```diff
@@ -0,0 +1,321 @@
+			default:
+				// doesn't support key other than "asc","desc",""
+				supportedOrders := []string{ascendantKey, descendantKey}
+				return gs, fmt.Errorf("sort order %q not supported, it must be in [%s]",
```
```go
return nil, fmt.Errorf(".....")
```

mstanleyjones replied:
not supported. It must be in...

(you need a period or semicolon, not a comma)

thaJeztah commented on cli/command/formatter/sorter.go:
```diff
@@ -0,0 +1,321 @@
+		case 1:
+			// sort order not specified, keep default
+		default:
+			return gs, fmt.Errorf("malformed sort key: %q", by)
```
```go
return nil, fmt.Errorf(".....")
```

thaJeztah commented on cli/command/formatter/sorter.go:
```diff
@@ -0,0 +1,321 @@
+	return 1
+}
+
+func getSortableFields(dataType reflect.Type, whitelist map[string]string, parent fieldIndex, fields map[string]fieldIndex) error {
```
Is there a reason `getSortableFields()` doesn't actually _return_ the sortable fields? I think it's confusing, and more messy, as we now have to initialize a variable before we call this function every time? Thinking of;

```go
func getSortableFields(dataType reflect.Type, whitelist map[string]string, parent fieldIndex) (map[string]fieldIndex, error)
```

WeiZhang555 replied:
Because this is a recursive function, we need to pass `fields` from top to bottom, and all the levels will filling the same `fields` param.

thaJeztah replied:
I guess it's still possible to `append` the results? https://play.golang.org/p/gYha5yTfrO

thaJeztah commented on cli/command/formatter/sorter.go:
```diff
@@ -0,0 +1,321 @@
+		by = strings.TrimSpace(by)
+		parts := strings.Split(by, ":")
+		dir := ascendantKey // ascendant key is default
+		switch len(parts) {
```
This logic looks a bit over-complicated, I think we can simplify a bit if we use `strings.SplitN()`, and just check the second value (I included some of other comments in the change);

```go
// newGenericStructSorter creates a sorter based on data and user specified configure
// whitelist is a map with struct field name as key and new name as value
func newGenericStructSorter(data []interface{}, bys []string, whitelist map[string]string) (genericStructSorter, error) {
	gs := genericStructSorter{}
	if len(data) == 0 || len(bys) == 0 {
		return gs, nil
	}
	if err := validateType(data); err != nil {
		return nil, err
	}
	// scan and get the sortable fields
	sortableFields := make(map[string]fieldIndex)
	if err := getSortableFields(reflect.TypeOf(data[0]), whitelist, nil, sortableFields); err != nil {
		return nil, fmt.Errorf("failed to get sortable fields: %v", err)
	}

	gs.data = data
	for _, by := range bys {
		by = strings.TrimSpace(by)
		parts := strings.SplitN(by, ":", 2)
		dir := ascendantKey // ascendant key is default
		if len(parts) == 2 {
			switch parts[1] {
			case descendantKey:
				dir = descendantKey
			case ascendantKey:
				dir = ascendantKey
			default:
				// doesn't support key other than "asc","desc",""
				supportedOrders := []string{ascendantKey, descendantKey}
				return nil, fmt.Errorf(`invalid sort order: %q. Valid options are "%s"`,
					parts[1], strings.Join(supportedOrders, `", "`))
			}
		}
		by = strings.ToLower(parts[0])
		index, ok := sortableFields[by]
		if !ok {
			var keys []string
			for k := range sortableFields {
				keys = append(keys, k)
			}
			return nil, fmt.Errorf(`unknown sort field: %q. Valid options are "%s"`, parts[0], strings.Join(keys, `", "`))
		}
		gs.bys = append(gs.bys, orderBy{
			name:     by,
			ordering: dir,
			index:    index,
		})

	}

	return gs, nil
}
```

diff;

```patch
diff --git a/cli/command/formatter/sorter.go b/cli/command/formatter/sorter.go
index 23e52ff..a281b71 100644
--- a/cli/command/formatter/sorter.go
+++ b/cli/command/formatter/sorter.go
@@ -56,43 +56,35 @@ func validateType(data []interface{}) error {
 // whitelist is a map with struct field name as key and new name as value
 func newGenericStructSorter(data []interface{}, bys []string, whitelist map[string]string) (genericStructSorter, error) {
        gs := genericStructSorter{}
-       if len(data) == 0 {
+       if len(data) == 0 || len(bys) == 0 {
                return gs, nil
        }
-       if len(bys) == 0 {
-               return gs, fmt.Errorf("bys cannot be empty")
-       }
        if err := validateType(data); err != nil {
-               return gs, err
+               return nil, err
        }
        // scan and get the sortable fields
        sortableFields := make(map[string]fieldIndex)
        if err := getSortableFields(reflect.TypeOf(data[0]), whitelist, nil, sortableFields); err != nil {
-               return gs, fmt.Errorf("can't get sortable fields: %v", err)
+               return nil, fmt.Errorf("failed to get sortable fields: %v", err)
        }

        gs.data = data
        for _, by := range bys {
                by = strings.TrimSpace(by)
-               parts := strings.Split(by, ":")
+               parts := strings.SplitN(by, ":", 2)
                dir := ascendantKey // ascendant key is default
-               switch len(parts) {
-               case 2:
+               if len(parts) == 2 {
                        switch parts[1] {
                        case descendantKey:
                                dir = descendantKey
-                       case ascendantKey, "":
-                               // if default, keep as ascendantKey
+                       case ascendantKey:
+                               dir = ascendantKey
                        default:
                                // doesn't support key other than "asc","desc",""
                                supportedOrders := []string{ascendantKey, descendantKey}
-                               return gs, fmt.Errorf("sort order %q not supported, it must be in [%s]",
-                                       parts[1], strings.Join(supportedOrders, ","))
+                               return nil, fmt.Errorf(`invalid sort order: %q. Valid options are "%s"`,
+                                       parts[1], strings.Join(supportedOrders, `", "`))
                        }
-               case 1:
-                       // sort order not specified, keep default
-               default:
-                       return gs, fmt.Errorf("malformed sort key: %q", by)
                }
                by = strings.ToLower(parts[0])
                index, ok := sortableFields[by]
@@ -101,7 +93,7 @@ func newGenericStructSorter(data []interface{}, bys []string, whitelist map[stri
                        for k := range sortableFields {
                                keys = append(keys, k)
                        }
-                       return gs, fmt.Errorf("can't sort by %q, it must be in [%s]", parts[0], strings.Join(keys, ","))
+                       return nil, fmt.Errorf(`unknown sort field: %q. Valid options are "%s"`, parts[0], strings.Join(keys, `", "`))
                }
                gs.bys = append(gs.bys, orderBy{
                        name:     by,
```

WeiZhang555 replied:
This should work, I'll made a modification. :+1:

thaJeztah commented on docs/reference/commandline/images.md:
````diff
@@ -343,3 +344,39 @@ b6fa739cedf5        committ                   latest
 {% endraw %}
 ```
+
+## Customizing display orders
````
Perhaps;

```Markdown
## Customize sort order
```

(orders (plural) has a different meaning)

mstanleyjones replied:
Make this a level-3 heading to keep it under Description.

thaJeztah reviewed (changes requested):
I'm temporarily moving this back to code review, because I think there's some changes that can be made

mstanleyjones commented on cli/command/formatter/sorter.go:
```diff
@@ -0,0 +1,321 @@
+
+type orderBy struct {
+	name     string     // order by name
+	ordering string     // asc or desc
```
Would it be weird to just make it a boolean, `asc`, that if set to `false` would sort descending?

WeiZhang555 replied:
@mstanleyjones, that's a good suggestion :+1:

mstanleyjones commented on cli/command/formatter/sorter_test.go:
```diff
@@ -0,0 +1,253 @@
+	whitelist := map[string]string{"B": ""}
+	gs, err := newGenericStructSorter(data, []string{"B:desc"}, whitelist)
+	if err != nil {
+		t.Fatal("failed to create sorter:", err)
```
I think you need a space.

mstanleyjones commented on cli/command/formatter/sorter_test.go:
```diff
@@ -0,0 +1,253 @@
+	whitelist := map[string]string{"S": ""}
+	gs, err := newGenericStructSorter(data, []string{"S"}, whitelist)
+	if err != nil {
+		t.Fatal("failed to create sorter:", err)
```
I think you need a space.

mstanleyjones commented on cli/command/formatter/sorter_test.go:
```diff
@@ -0,0 +1,253 @@
+	whitelist := map[string]string{"I": ""}
+	gs, err := newGenericStructSorter(data, []string{"I:asc"}, whitelist)
+	if err != nil {
+		t.Fatal("failed to create sorter:", err)
```
space

mstanleyjones commented on cli/command/formatter/sorter_test.go:
```diff
@@ -0,0 +1,253 @@
+	whitelist := map[string]string{"U": ""}
+	gs, err := newGenericStructSorter(data, []string{"U"}, whitelist)
+	if err != nil {
+		t.Fatal("failed to create sorter:", err)
```
space

mstanleyjones commented on cli/command/formatter/sorter_test.go:
```diff
@@ -0,0 +1,253 @@
+	whitelist := map[string]string{"F": ""}
+	gs, err := newGenericStructSorter(data, []string{"F"}, whitelist)
+	if err != nil {
+		t.Fatal("failed to create sorter:", err)
```
space

mstanleyjones commented on cli/command/image/list.go:
```diff
@@ -46,6 +47,7 @@ func NewImagesCommand(dockerCli *command.DockerCli) *cobra.Command {
 	flags.BoolVar(&opts.showDigests, "digests", false, "Show digests")
 	flags.StringVar(&opts.format, "format", "", "Pretty-print images using a Go template")
 	flags.VarP(&opts.filter, "filter", "f", "Filter output based on conditions provided")
+	flags.StringVar(&opts.sortby, "sort-by", "", "Customize list order, sort by field[:order]. e.g. \"id:asc,createdat:desc\"")
```
Try to avoid "e.g." and other Latin abbreviations. "For example" is fine.

WeiZhang555 replied:
Won't that be too long for a help message?

mstanleyjones commented on docs/reference/commandline/images.md:
```diff
@@ -33,6 +33,7 @@ Options:
       --help            Print usage
       --no-trunc        Don't truncate output
   -q, --quiet           Only show numeric IDs
+      --sort-by string  Customize list order, sort by field[:order]. e.g. "id:asc,createdat:desc"
```
If you change the message above, sync here.

mstanleyjones commented on docs/reference/commandline/images.md:
```diff
@@ -343,3 +344,39 @@ b6fa739cedf5        committ                   latest
+
+## Customizing display orders
+
+To customize list orders, you can use `--sort-by` option, this option takes format
```
option, which takes the format

(for "of" below, you need "the" here)

mstanleyjones commented on docs/reference/commandline/images.md:
```diff
@@ -343,3 +344,39 @@ b6fa739cedf5        committ                   latest
+## Customizing display orders
+
+To customize list orders, you can use `--sort-by` option, this option takes format
+of `field[:order]`. `field` should be a case insensitive string among "repository",
```
s/should be/is

How about `field` is one or more of: `repository`, `tag`, `digest`, `createdat`, `id`, or `size`.

(I assume you can sort by more than one)

WeiZhang555 replied:
Yes, we can sort by more than one.

mstanleyjones commented on docs/reference/commandline/images.md:
```diff
@@ -343,3 +344,39 @@ b6fa739cedf5        committ                   latest
+
+To customize list orders, you can use `--sort-by` option, this option takes format
+of `field[:order]`. `field` should be a case insensitive string among "repository",
+"tag", "digest", "createdat", "id" and "size". `order` should be either `asc`(ascendant)
```
`order` is either `asc` (ascending) or `desc` (descending).

mstanleyjones commented on docs/reference/commandline/images.md:
```diff
@@ -343,3 +344,39 @@ b6fa739cedf5        committ                   latest
+of `field[:order]`. `field` should be a case insensitive string among "repository",
+"tag", "digest", "createdat", "id" and "size". `order` should be either `asc`(ascendant)
+or `desc`(descendant).
+The meaning of each `field` is described in [`Formatting`](#formatting) section.
```
described in the ...

mstanleyjones commented on docs/reference/commandline/images.md:
```diff
@@ -343,3 +344,39 @@ b6fa739cedf5        committ                   latest
+or `desc`(descendant).
+The meaning of each `field` is described in [`Formatting`](#formatting) section.
+
+If you specify a wrong `field` or `order` string, it will output the right options
```
s/a wrong/an invalid

s/it will output...options/a hint will be printed, such as the following:

mstanleyjones commented on docs/reference/commandline/images.md:
````diff
@@ -343,3 +344,39 @@ b6fa739cedf5        committ                   latest
+as a hint, like:
+
+```bash
+{% raw %}
````
you don't need the raw tags, only if you have double brackets.

mstanleyjones commented on integration-cli/docker_cli_images_test.go:
```diff
@@ -363,3 +363,20 @@ func (s *DockerSuite) TestImagesFormatDefaultFormat(c *check.C) {
+	dockerCmd(c, "tag", "busybox", "name1:v2")
+	dockerCmd(c, "tag", "busybox", "name0:v2")
+
+	imagesOut, _ := dockerCmd(c, "images", "--sort-by", "Repository:asc,tag:desc", "name*")
```
why is this different from line 379?

WeiZhang555 replied:
Because this is the correct usage, as a compare, line 379 demonstrate invalid options.

WeiZhang555 commented on cli/command/formatter/sorter.go:
```diff
@@ -0,0 +1,321 @@
+				// doesn't support key other than "asc","desc",""
+				supportedOrders := []string{ascendantKey, descendantKey}
+				return gs, fmt.Errorf("sort order %q not supported, it must be in [%s]",
+					parts[1], strings.Join(supportedOrders, ","))
```
OK

WeiZhang555 commented:
@thaJeztah @mstanleyjones All your comments are addressed.

mstanleyjones commented on cli/command/formatter/sorter.go:
```diff
@@ -0,0 +1,321 @@
+	"time"
+)
+
+const (
```
If you make it boolean as suggested elsewhere, you don't need this either

WeiZhang555 replied:
I think this is a little bit different, this is a keyword, so it has to be string.

It is supposed to appear in `--sort-by createdat:asc` or `--sort-by repository:desc`, so here we can't convert it to bool.

mstanleyjones commented on docs/reference/commandline/images.md:
````diff
@@ -343,3 +344,34 @@ b6fa739cedf5        committ                   latest
+
+```bash
+$ docker images --sort-by invalid
+can't sort by "invalid", it must be in [repository,tag,digest,createdat,id,size]
````
Try to avoid using contractions in error messages. It's a little too informal and it's also really annoying to document because Markdown sometimes has trouble differentiating an apostrophe from an unclosed single quote.

WeiZhang555 replied:
Ooops, I forgot to update this. I've updated error message in codes but forget to update related docs. 

Updating now...

WeiZhang555 committed:
Refactor sorter to make codes cleaner

Signed-off-by: Zhang Wei <zhangwei555@huawei.com>

mdlinville reviewed (approved):
Docs LGTM

WeiZhang555 commented:
@thaJeztah What do you think about this?

WeiZhang555 commented:
@thaJeztah so are we waiting decision of https://github.com/docker/docker/issues/31062 ? Where are we now? :smile:

AkihiroSuda commented:
@thaJeztah PTAL?

WeiZhang555 commented:
ping @thaJeztah , any news?

thaJeztah commented:
@WeiZhang555 sorry for the long delay; we actually discussed this PR during a maintainers meeting, and there were some concerns about this related to https://github.com/docker/docker/issues/31062; I checked yesterday, and @vieux sent you an e-mail (hope the e-mail address you're using in your sign-offs is correct, otherwise feel free to mail me at sebastiaan at docker.com if you didn't receive anything

WeiZhang555 commented:
Based on information I got, we're planning to go forward with #31062, and it has some conflicts with this PR from design point of view, so I'm closing this, and let's focus on #31062 .

Also thank for everyone for help me review this :smile:
